### ISSUE TITLE
25.3.0-beta2 fails to build on Linux

### ISSUE THREAD

**Issue by WorMzy** [MoonchildProductions/Pale-Moon#19]
Hi, I'm getting a build failure with beta2:

```
 8:47.34 GLScreenBuffer.cpp
 8:47.37 /build/palemoon-beta/src/Pale-Moon/gfx/gl/GLContextProviderGLX.cpp: In member function 'bool mozilla::gl::GLXLibrary::EnsureInitialized(mozilla::gl::LibType)':
 8:47.37 /build/palemoon-beta/src/Pale-Moon/gfx/gl/GLContextProviderGLX.cpp:112:9: error: 'ScopedGfxFeatureReporter' was not declared in this scope
 8:47.37          ScopedGfxFeatureReporter reporter(libGLfilename, forceFeatureReport);
 8:47.37          ^
 8:47.37 /build/palemoon-beta/src/Pale-Moon/gfx/gl/GLContextProviderGLX.cpp:118:9: error: 'reporter' was not declared in this scope
 8:47.37          reporter.SetSuccessful();
 8:47.37          ^
 8:47.42 GLTextureImage.cpp
 8:47.47 SharedSurface.cpp
 8:47.50 
 8:47.50 In the directory  /build/palemoon-beta/src/pmbuild/gfx/gl
 8:47.50 The following command failed to execute properly:
 8:47.50 g++ -o GLContextProviderGLX.o -c -I../../dist/stl_wrappers -I../../dist/system_wrappers -include /build/palemoon-beta/src/Pale-Moon/config/gcc_hidden.h -DMOZ_GLUE_IN_PROGRAM -DMOZILLA_INTERNAL_API -D_IMPL_NS_COM -DEXPORT_XPT_API -DEXPORT_XPTC_API -D_IMP$
_NS_GFX -D_IMPL_NS_WIDGET -DIMPL_XREAPI -DIMPL_NS_NET -DIMPL_THEBES -DSTATIC_EXPORTABLE_JS_API -DNO_NSPR_10_SUPPORT -DOS_POSIX=1 -DOS_LINUX=1 -I/build/palemoon-beta/src/Pale-Moon/ipc/chromium/src -I/build/palemoon-beta/src/Pale-Moon/ipc/glue -I../../ipc/ipdl/_ip$
lheaders -I/build/palemoon-beta/src/Pale-Moon/gfx/gl -I. -I../../dist/include -I/build/palemoon-beta/src/pmbuild/dist/include/nspr -I/build/palemoon-beta/src/pmbuild/dist/include/nss -fPIC -D_FORTIFY_SOURCE=2 -O2 -Wall -Wpointer-arith -Woverloaded-virtual -Werro$
=return-type -Wtype-limits -Wempty-body -Wsign-compare -Wno-invalid-offsetof -Wcast-align -march=x86-64 -mtune=native -O2 -pipe -fstack-protector-strong --param=ssp-buffer-size=4 -fno-exceptions -fno-strict-aliasing -fno-rtti -ffunction-sections -fdata-sections $
fno-exceptions -std=gnu++0x -pthread -pipe -DNDEBUG -DTRIMMED -Os -freorder-blocks -fomit-frame-pointer -I/build/palemoon-beta/src/pmbuild/dist/include/cairo -I/build/palemoon-beta/src/Pale-Moon/widget/gtk2/compat -pthread -I/usr/include/gtk-2.0 -I/usr/include/g$
k-unix-print-2.0 -I/usr/include/gtk-2.0 -I/usr/include/atk-1.0 -I/usr/include/pango-1.0 -I/usr/include/gtk-2.0 -I/usr/lib/gtk-2.0/include -I/usr/include/pango-1.0 -I/usr/include/cairo -I/usr/include/pixman-1 -I/usr/include/freetype2 -I/usr/include/libpng16 -I/us$
/include/harfbuzz -I/usr/include/freetype2 -I/usr/include/harfbuzz -I/usr/include/libdrm -I/usr/include/libpng16 -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/libpng16 -I/usr/include/glib-2.0 -I/usr/lib/glib-2.0/include -D_FORTIFY_SOURCE=2 -O2 -DMOZILLA_CLIENT -i$
clude ../../mozilla-config.h -MD -MP -MF .deps/GLContextProviderGLX.o.pp /build/palemoon-beta/src/Pale-Moon/gfx/gl/GLContextProviderGLX.cpp
 8:47.50 /build/palemoon-beta/src/Pale-Moon/config/rules.mk:1104: recipe for target 'GLContextProviderGLX.o' failed
 8:47.50 make[6]: *** [GLContextProviderGLX.o] Error 1
 8:47.50 make[6]: *** Waiting for unfinished jobs....
 8:49.23 /build/palemoon-beta/src/Pale-Moon/config/makefiles/target_libs.mk:59: recipe for target 'libs' failed
 8:49.23 make[5]: *** [libs] Error 2
 8:49.23 /build/palemoon-beta/src/Pale-Moon/config/makefiles/target_libs.mk:16: recipe for target 'libs_tier_platform' failed
 8:49.23 make[4]: *** [libs_tier_platform] Error 2
 8:49.23 /build/palemoon-beta/src/Pale-Moon/config/rules.mk:723: recipe for target 'tier_platform' failed
 8:49.23 make[3]: *** [tier_platform] Error 2
 8:49.23 /build/palemoon-beta/src/Pale-Moon/config/rules.mk:656: recipe for target 'default' failed
 8:49.23 make[2]: *** [default] Error 2
 8:49.23 /build/palemoon-beta/src/Pale-Moon/client.mk:372: recipe for target 'realbuild' failed
 8:49.23 make[1]: *** [realbuild] Error 2
 8:49.23 client.mk:172: recipe for target 'build' failed
 8:49.23 make: *** [build] Error 2
 8:49.24 314 compiler warnings present.
```
For the record, beta1 built fine. I'll try to find time to bisect this tomorrow (today here) if it will be helpful.

**Comment by wolfbeast** [comment]
Thanks for reporting!
Please remove both reported statements from that file which would allow you to build; I'll get a patch in for it - it's no longer defined with the removal of the gfxinfo reporter (part of the crash reporting system). That one slipped through the cracks ;)

**Comment by WorMzy** [comment]
Hi, thanks for the quick reply!

Unfortunately, after removing those two lines, it uncovered a lot more errors:

```
26:29.37 nsIUpdateTimerManager.idl
26:29.46 defined(@array) is deprecated at /build/palemoon-beta/src/Pale-Moon/config/milestone.pl line 59.
26:29.46        (Maybe you should just omit the defined()?)
26:29.47 nsNativeAppSupportUnix.cpp
26:29.48 defined(@array) is deprecated at /build/palemoon-beta/src/Pale-Moon/config/milestone.pl line 59.
26:29.48        (Maybe you should just omit the defined()?)
26:29.48 nsX11ErrorHandler.cpp
26:29.50 defined(@array) is deprecated at /build/palemoon-beta/src/Pale-Moon/config/milestone.pl line 59.
26:29.50        (Maybe you should just omit the defined()?)
26:29.50 CreateAppData.cpp
26:29.51 defined(@array) is deprecated at /build/palemoon-beta/src/Pale-Moon/config/milestone.pl line 59.
26:29.51        (Maybe you should just omit the defined()?)
26:29.51 ProfileReset.cpp
26:29.53 defined(@array) is deprecated at /build/palemoon-beta/src/Pale-Moon/config/milestone.pl line 59.
26:29.53        (Maybe you should just omit the defined()?)
26:29.53 nsAppRunner.cpp
26:29.56 defined(@array) is deprecated at /build/palemoon-beta/src/Pale-Moon/config/milestone.pl line 59.
26:29.56        (Maybe you should just omit the defined()?)
26:29.72 /build/palemoon-beta/src/Pale-Moon/toolkit/xre/nsX11ErrorHandler.cpp: In function 'int X11Error(Display*, XErrorEvent*)':
26:29.72 /build/palemoon-beta/src/Pale-Moon/toolkit/xre/nsX11ErrorHandler.cpp:25:3: error: 'nsAutoCString' was not declared in this scope
26:29.72    nsAutoCString message;
26:29.72    ^
26:29.72 /build/palemoon-beta/src/Pale-Moon/toolkit/xre/nsX11ErrorHandler.cpp:28:5: error: 'message' was not declared in this scope
26:29.72      message.AppendInt(event->request_code);
26:29.72      ^
26:29.72 /build/palemoon-beta/src/Pale-Moon/toolkit/xre/nsX11ErrorHandler.cpp:49:13: error: 'message' was not declared in this scope
26:29.72              message.Append(extNames[i]);
26:29.72              ^
26:29.72 /build/palemoon-beta/src/Pale-Moon/toolkit/xre/nsX11ErrorHandler.cpp:65:11: error: 'message' was not declared in this scope
26:29.72        if (message.EqualsLiteral("XInputExtension.4") &&
26:29.72            ^
26:29.72 /build/palemoon-beta/src/Pale-Moon/toolkit/xre/nsX11ErrorHandler.cpp:74:7: error: 'message' was not declared in this scope
26:29.72    if (message.IsEmpty()) {
26:29.72        ^
26:29.73 /build/palemoon-beta/src/Pale-Moon/toolkit/xre/nsX11ErrorHandler.cpp:81:17: error: expected ';' before 'notes'
26:29.73    nsAutoCString notes;
26:29.73                  ^
26:29.73 /build/palemoon-beta/src/Pale-Moon/toolkit/xre/nsX11ErrorHandler.cpp:83:5: error: 'notes' was not declared in this scope
26:29.73      notes.Append(buffer);
26:29.73      ^
26:29.73 /build/palemoon-beta/src/Pale-Moon/toolkit/xre/nsX11ErrorHandler.cpp:85:5: error: 'notes' was not declared in this scope
26:29.73      notes.Append("Request ");
26:29.73      ^
26:29.73 /build/palemoon-beta/src/Pale-Moon/toolkit/xre/nsX11ErrorHandler.cpp:91:3: error: 'notes' was not declared in this scope
26:29.73    notes.Append(": ");
26:29.73    ^
26:29.73 
26:29.73 In the directory  /build/palemoon-beta/src/pmbuild/toolkit/xre
26:29.73 The following command failed to execute properly:
26:29.73 g++ -o nsX11ErrorHandler.o -c -I../../dist/stl_wrappers -I../../dist/system_wrappers -include /build/palemoon-beta/src/Pale-Moon/config/gcc_hidden.h -DUSE_GLX_TEST -DIMPL_XREAPI -DMOZ_APP_NAME="palemoon" -DMOZ_APP_VERSION="25.3.0b2" -DMOZ_GLUE_IN_PROGRAM -DMOZILLA_INTERNAL
_API -D_IMPL_NS_COM -DEXPORT_XPT_API -DEXPORT_XPTC_API -D_IMPL_NS_GFX -D_IMPL_NS_WIDGET -DIMPL_XREAPI -DIMPL_NS_NET -DIMPL_THEBES -DSTATIC_EXPORTABLE_JS_API -DNO_NSPR_10_SUPPORT -DOS_POSIX=1 -DOS_LINUX=1 -DOS_TARGET="Linux" -DMOZ_WIDGET_TOOLKIT="gtk2" -DTARGET_XPCOM_ABI="x86_64-gcc3" -DTARGET_OS_ABI="Linux_x86_64-gcc3" -DTOOLKIT_EM_VERSION="25.3.0" -DWRAP_SYSTEM_INCLUDES -DGRE_MILESTONE=25.3.0 -DGRE_BUILDID=20150205095747 -DMOZILLA_OFFICIAL -DAPP_VERSION=25.3.0b2 -DAPP_ID={8de7fcbb-c55c-4fbe-bfc5-fc555c87dbc4} -I/build/palemoon-beta/src/Pale-Moon/widget/xremoteclient -I/build/palemoon-beta/src/Pale-Moon/ipc/chromium/src -I/build/palemoon-beta/src/Pale-Moon/ipc/glue -I../../ipc/ipdl/_ipdlheaders -I/build/palemoon-beta/src/Pale-Moon/testing/gtest/mozilla -I/build/palemoon-beta/src/Pale-Moon/toolkit/xre/../profile -I/build/palemoon-beta/src/Pale-Moon/dom/ipc -I/build/palemoon-beta/src/Pale-Moon/dom/base -I/build/palemoon-beta/src/Pale-Moon/xpcom/build -I/build/palemoon-beta/src/Pale-Moon/config -I/build/palemoon-beta/src/Pale-Moon/toolkit/xre -I. -I../../dist/include -I/build/palemoon-beta/src/pmbuild/dist/include
/nspr -I/build/palemoon-beta/src/pmbuild/dist/include/nss -fPIC -D_FORTIFY_SOURCE=2 -O2 -Wall -Wpointer-arith -Woverloaded-virtual -Werror=return-type -Wtype-limits -Wempty-body -Wsign-compare -Wno-invalid-offsetof -Wcast-align -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong --param=ssp-buffer-size=4 -fno-exceptions -fno-strict-aliasing -fno-rtti -ffunction-sections -fdata-sections -fno-exceptions -std=gnu++0x -pthread -pipe -DNDEBUG -DTRIMMED -Os -freorder-blocks -fomit-frame-pointer -I/build/palemoon-beta/src/Pale-Moon/widget/gtk2/compat -pthread -I/usr/include/gtk-2.0 -I/usr/include/gtk-unix-print-2.0 -I/usr/include/gtk-2.0 -I/usr/include/atk-1.0 -I/usr/include/pango-1.0 -I/usr/include/gtk-2.0 -I/usr/lib/gtk-2.0/include -I/usr/include/pango-1.0 -I/usr/include/cairo -I/usr/include/pixman-1 -I/usr/include/freetype2 -I
/usr/include/libpng16 -I/usr/include/harfbuzz -I/usr/include/freetype2 -I/usr/include/harfbuzz -I/usr/include/libdrm -I/usr/include/libpng16 -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/libpng16 -I/usr/include/glib-2.0 -I/usr/lib/glib-2.0/include -I/usr/include/dbus-1.0 -I/usr/lib/dbus-1.0/include -I/usr/include/dbus-1.0 -I/usr/lib/dbus-1.0/include -I/usr/include/glib-2.0 -I/usr/lib/glib-2.0/include -pthread -I/usr/include/pango-1.0 -I/usr/include/cairo -I/usr/include/glib-2.0 -I/usr/lib/glib-2.0/include -I/usr/include/pixman-1 -I/usr/include/freetype2 -I/us
r/include/libpng16 -I/usr/include/harfbuzz -I/usr/include/freetype2 -I/usr/include/harfbuzz -I/usr/include/libdrm -I/usr/include/libpng16 -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/harfbuzz -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/harfbuz
z -D_FORTIFY_SOURCE=2 -O2 -DMOZILLA_CLIENT -include ../../mozilla-config.h -MD -MP -MF .deps/nsX11ErrorHandler.o.pp /build/palemoon-beta/src/Pale-Moon/toolkit/xre/nsX11ErrorHandler.cpp
26:29.73 /build/palemoon-beta/src/Pale-Moon/config/rules.mk:1104: recipe for target 'nsX11ErrorHandler.o' failed
26:29.73 make[6]: *** [nsX11ErrorHandler.o] Error 1
26:29.73 make[6]: *** Waiting for unfinished jobs....
26:30.55 /build/palemoon-beta/src/Pale-Moon/config/makefiles/target_libs.mk:59: recipe for target 'libs' failed
26:30.55 make[5]: *** [libs] Error 2
26:30.55 /build/palemoon-beta/src/Pale-Moon/config/makefiles/target_libs.mk:16: recipe for target 'libs_tier_platform' failed
26:30.55 make[4]: *** [libs_tier_platform] Error 2
26:30.55 /build/palemoon-beta/src/Pale-Moon/config/rules.mk:723: recipe for target 'tier_platform' failed
26:30.55 make[3]: *** [tier_platform] Error 2
26:30.55 /build/palemoon-beta/src/Pale-Moon/config/rules.mk:656: recipe for target 'default' failed
26:30.55 make[2]: *** [default] Error 2
26:30.55 /build/palemoon-beta/src/Pale-Moon/client.mk:372: recipe for target 'realbuild' failed
26:30.55 make[1]: *** [realbuild] Error 2
26:30.55 client.mk:172: recipe for target 'build' failed
26:30.55 make: *** [build] Error 2
26:30.59 565 compiler warnings present.
```

I've hollowed out the X11Error function (leaving just 'return 0;'), and started a rebuild, so I'll see if there are any additional problems lurking.

**Comment by WorMzy** [comment]
It built successfully after that, although I'm not sure what it would do in the event of an X11 error..

**Comment by wolfbeast** [comment]
I'll have a look. I've only built on Windows so far and haven't had a chance to poke the linux dev about it, so yes, it's a beta stage, and no, it might not actually build out of the box after a massive commit with >38k changes. We don't have the infrastructure to have multiple try branches or always build every platform in each stage of development.

**Comment by WorMzy** [comment]
That's fine, I know that Linux isn't your primary concern, and breakage will happen occasionally. If I knew what I was doing, I'd take a stab at fixing it and sending a pull request. Hopefully my reports help either yourself or the linux dev in some way. :)

**Comment by wolfbeast** [comment]
I think it's just missing **nsString.h** in the **#includes**, actually, since that was previously linked through the (now removed) exceptionhandler.h and that's your error (undefined nsAutoCString type)
I'm not entirely sure if X11Error() is optional, and it's better to be safe and leave it functional even if some built strings are never used in the end - I'll have to investigate further but X11 isn't my forté ;)

**Comment by wolfbeast** [comment]
Fixed by commit d6fef27b4b0f2294572f00ed845804b353d4ddef

**Comment by WorMzy** [comment]
Thanks, I can confirm that this commit fixes the build failure.
